src/matrix/mod.rs

```diff
@@ -95,114 +115,50 @@ impl MatrixClient {
         &mut self,
-        room_id: &str,
+        room_id: identifiers::RoomId,
         body: String,
-    ) -> Result<protocol::RoomSendResponse, Error> {
-        let msg_id = thread_rng().gen::<u16>();
-        let mut url = self
-            .url
-            .join(&format!(
-                "/_matrix/client/r0/rooms/{}/send/m.room.message/mircd-{}",
-                room_id, msg_id
-            ))
-            .expect("Unable to construct a valid API url");
-
-        url.query_pairs_mut()
-            .clear()
-            .append_pair("access_token", &self.access_token);
-
-        let body = serde_json::to_vec(&protocol::RoomSendInput {
-            body,
-            msgtype: "m.text".to_string(),
-        })?
-        .into();
-
-        let request = hyper::Request::builder()
-            .method("PUT")
-            .uri(url.as_str().parse::<hyper::Uri>().unwrap())
-            .body(body)
-            .expect("request could not be built");
+    ) -> Result<api::message::create_message_event::Response, Error> {
+        let txn_id= thread_rng().gen::<u16>().to_string();
+        let event_type = events::EventType::RoomMessage;
+        let data = TextMessage::new(body).raw_json()?;
+
+        let request = api::message::create_message_event::Request {
+            room_id,
+            event_type,
+            txn_id,
+            data
+        };
 
-        // TODO: improve this error handling
-        let response = self.http_client.send_request(request).await?;
-        let bytes = hyper::body::to_bytes(response.into_body()).await?;
+        let response = self.client.request(request).await?;
 
-        let json_response = serde_json::from_slice(&bytes)?;
-        Ok(json_response)
+        Ok(response)
     }
 
-    pub async fn join_room(&mut self, room_id: &str) -> Result<protocol::RoomJoinResponse, Error> {
-        let roomid_encoded = percent_encode(room_id.as_bytes(), PATH_SEGMENT_ENCODE_SET);
-        let mut url = self
-            .url
-            .join(&format!("/_matrix/client/r0/join/{}", roomid_encoded))
-            .expect("Unable to construct a valid API url");
-
-        url.query_pairs_mut()
-            .clear()
-            .append_pair("access_token", &self.access_token);
-
-        let json_bytes = serde_json::to_vec(&protocol::RoomJoinInput {})?;
-
-        let request = hyper::Request::builder()
-            .method("POST")
-            .uri(url.as_str().parse::<hyper::Uri>().unwrap())
-            .body(json_bytes.into())
-            .expect("request could not be built");
-
-        let response = self.http_client.send_request(request).await?;
-        let bytes = hyper::body::to_bytes(response.into_body()).await?;
-        let json_response = serde_json::from_slice(&bytes)?;
-        Ok(json_response)
+    pub async fn join_room(&mut self, room_id: identifiers::RoomId) -> Result<api::membership::join_room_by_id::Response, Error> {
+        let request = api::membership::join_room_by_id::Request {
```

Comment:
The endpoint that was previously used here is a different one.

```suggestion
        let request = api::membership::join_room_by_id_or_alias::Request {
```

Reply:
Fixed